packages/perseus/src/util/parse-perseus-json/perseus-parsers/versioned-widget-options.ts

```diff
@@ -13,75 +21,93 @@ type Versioned = {
  * data does not match any of the versions.
  *
  * @example
- * const parseOptions = versionedWidgetOptions(parseOptionsV3)
- *     .withMigrationFrom(parseOptionsV2, migrateV2ToV3)
- *     .withMigrationFrom(parseOptionsV1, migrateV1ToV2)
- *     .withMigrationFrom(parseOptionsV0, migrateV0ToV1)
+ * const parseOptions = versionedWidgetOptions(3, parseOptionsV3)
+ *     .withMigrationFrom(2, parseOptionsV2, migrateV2ToV3)
+ *     .withMigrationFrom(1, parseOptionsV1, migrateV1ToV2)
+ *     .withMigrationFrom(0, parseOptionsV0, migrateV0ToV1)
  *     .parser;
  *
+ * @param latestMajorVersion the latest major version of the widget options.
  * @param parseLatest a {@link Parser} for the latest version of the widget
- * options. This should check version.major and fail if it's not the latest
- * version.
+ * options.
  * @returns a builder object, to which migrations from earlier versions can be
  * added. Migrations must be added in "reverse chronological" order as in the
  * example above.
  */
 export function versionedWidgetOptions<Latest extends Versioned>(
+    latestMajorVersion: number,
     parseLatest: Parser<Latest>,
 ): VersionedWidgetOptionsParserBuilder<Latest, Latest> {
     return new VersionedWidgetOptionsParserBuilder(
+        latestMajorVersion,
         parseLatest,
         (latest) => latest,
+        (raw, ctx) =>
+            ctx.failure("widget options with a known version number", raw),
     );
 }
 
 class VersionedWidgetOptionsParserBuilder<
+    // The latest version of the widget options
     Latest extends Versioned,
-    Migratable extends Versioned,
+    // A version of the widget options which we want to migrate to Latest.
+    MigratableWidgetOptions extends Versioned,
 > {
+    public parser: Parser<Latest>;
+
     constructor(
-        public parser: Parser<Latest>,
-        private migrate: (m: Migratable) => Latest,
-    ) {}
+        majorVersion: number,
+        parseThisVersion: Parser<MigratableWidgetOptions>,
+        private migrateToLatest: (m: MigratableWidgetOptions) => Latest,
+        private parseOtherVersions: Parser<Latest>,
```

Comment:
Wait! What? Does this make these two constructor parameters into private class members and assign them automatically?  I've never seen this but if that's what it does, that's pretty awesome. :til:

Reply:
Yep! :)